Add siteSlug prop to CompositeCheckoutContainer

nbloomf opened the pull request:
#### Changes proposed in this Pull Request

* All control paths into the `checkout` middleware, where new checkout will be rendered, will have already created a cart on the server. This means that to get the canonical cart data that we need it's enough to call `getCart` with the cart slug.

We also specialize the `setCart` handler to fix the key parameter.

To do:
- [x] Rebase after #37977
- [x] Merge in Filippo's changes from #37941
- [x] Get removeItem to work

#### Testing instructions
Automated tests: `npm run test-packages composite-checkout-wpcom`

Manual test:
- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`
- Go through checkout flow for a new site. The selected products should load in the checkout form.

Commit:
Pass live shopping cart endpoint wrappers into composite checkout

sirbrillig commented:
Great work identifying the types of carts and typing this code!

fditrapani committed:
Composite checkout: Add delete to modal

sirbrillig commented:
This will probably have a couple of conflicts with #37937 but I think they're relatively minor.

Commit:
Incorporate feedback

sirbrillig commented on client/my-sites/checkout/checkout/composite-checkout-container.js:
```diff
@@ -19,52 +19,19 @@ import {
-);
+export const getServerCart = wpcom.getCart;
+
+const useShoppingCart = cartKey => {
```
I think it might be best not to start this function with `use` since it's not a hook, it's a hook creator creator. Maybe something like `buildMakeShoppingCartHook`? 🤔 Although that's pretty repetative. Is it possible to pass `cartKey` as an argument to `makeShoppingCartHook` directly?

This also makes me wonder if it's safe to be re-creating a hook on every render. With the old version of `makeShoppingCartHook`, it was called when the file loaded so there was only ever one version of the actual hook. This version will call the creator creator on every render, which will create a new version of the hook. If the hook has any internal state, I think this might cause problems.

So a better question might be: since `cartKey` is dynamic and changes on render (and is therefore a dependency for any effects of the hook), is it possible to pass `cartKey` as an argument to the hook itself, rather than preparing it here?

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-checkout-order-review.js:
```diff
@@ -18,7 +18,7 @@ import {
 } from './wp-order-review-line-items';
 
-export default function WPCheckoutOrderReview( { className } ) {
+export function WPCheckoutOrderReview( { className, removeItem } ) {
```
I don't mind, but is there a reason to change this to a named export rather than the default? I don't see other exports...

sirbrillig commented on packages/composite-checkout-wpcom/src/components/wp-order-review-line-items.js:
```diff
@@ -32,12 +31,12 @@ const OrderReviewSectionArea = styled.div`
 	const itemSpanId = `checkout-line-item-${ item.id }`;
 	const deleteButtonId = `checkout-delete-button-${ item.id }`;
 	const [ isModalVisible, setIsModalVisible ] = useState( false );
+	const hasDomainsInCart = useHasDomainsInCart();
```
Super minor, but is there any reason for moving this down, or was it just an artifact of refactoring? (In a big PR it's nice to avoid any unnecessary changes to make it easier to review.)

nbloomf replied:
It's an artifact- thanks for the heads up.

sirbrillig commented on packages/composite-checkout-wpcom/src/types.ts:
```diff
@@ -6,24 +6,39 @@ import {
+	RequestCartProduct,
+	ResponseCart,
+	ResponseCartProduct,
+    emptyResponseCart,
```
Super minor: indent is different on this line.

sirbrillig commented on packages/composite-checkout-wpcom/src/types.ts:
```diff
@@ -6,24 +6,39 @@ import {
+	RequestCartProduct,
+	ResponseCart,
+	ResponseCartProduct,
+    emptyResponseCart,
```
Different indent again.

sirbrillig committed:
Make useShoppingCart a static hook

useShoppingCart was being dynamically created by the
makeShoppingCartHook function. There's nothing wrong with this in
theory, but it does introduce additional complexity and might make the
code harder to understand since Hooks are stateful functions with
side-effects by definition.

This commit changes the code such that useShoppingCart is a regular Hook
instead.

One advantage of this is that the data dependencies of the hook become
more explicit. It requires functions to call the API (we could import
those directly into the Hook file instead but this way they can be
controlled by the main host component where all the other host-specific
logic lives; either way would be fine) and a key which in theory may
change (but in practice should not).

Commit:
Fixing test

sirbrillig committed:
Add debug to state changes

Commit:
Fix race condition in cart fetching

Commit:
Fix lint errors

Commit:
Fix lint errors

sirbrillig commented on client/my-sites/checkout/checkout/composite-checkout-container.js:
```diff
@@ -125,7 +79,10 @@ export function isApplePayAvailable() {
 const availablePaymentMethods = [ applePayMethod, stripeMethod, paypalMethod ].filter( Boolean );
 
-export default function CompositeCheckoutContainer() {
+const getCart = ( ...args ) => wpcom.getCart( ...args );
```
I know I was the one to add these, but we should check to see if this is required or if we can directly pass `wpcom.getCart`/`wpcom.setCart` to the props in `WPCheckoutWrapper`. If we do need them, as I suspect, we should add a comment above this line to say something like "wpcom functions require binding to the wpcom object to function correctly".

nbloomf replied:
I think this is required - passing `wpcom.getCart` and `wpcom.setCart` directly gives `TypeError: "this is undefined"`.

sirbrillig commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -22,7 +22,7 @@ import { isGSuiteRestricted } from 'lib/gsuite';
 import { sites } from 'my-sites/controller';
 import config from 'config';
-import CompositeCheckoutContainer from './checkout/composite-checkout-container';
+import { CompositeCheckoutContainer } from './checkout/composite-checkout-container';
```
`CompositeCheckoutContainer` is the default export of its file, so I don't think this should be working 🤔

sirbrillig commented on packages/composite-checkout-wpcom/src/index.js:
```diff
@@ -2,9 +2,15 @@
-export { WPCheckoutWrapper, makeShoppingCartHook, mockCartEndpoint, mockPayPalExpressRequest };
+export {
+	WPCheckoutWrapper,
+	makeShoppingCartHook,
```
I think this no longer exists and is now `useShoppingCart`

sirbrillig commented:
This also needs testing instructions.

sirbrillig committed:
Fix CompositeCheckoutContainer import

sirbrillig committed:
Fix props of CompositeCheckoutContainer

sirbrillig committed:
Convert index to string in removeItem

Commit:
Rename shopping-cart-hook to use-shopping-cart

Commit:
Add comment explaining bindings to wpcom object

sirbrillig reviewed (approved):
Code looks good to me. I haven't tested this last push manually but if you've done so, I'd consider this good to go. Testing before the last two commits worked great.
